warn/warn_docstring.go

```diff
@@ -0,0 +1,250 @@
+		// Exactly one line in the first block
+		info.hasHeader = true
+	}
+
+	// Iterate over the blocks, extract data
+	for _, block := range blocks {
+		switch block.lines[0] {
+		case "Args:", "Arguments:":
+			if block.lines[0] == "Arguments:" {
+				info.argumentsLine = block.startLineNo
+			}
+
+			argIndentation := -1
+			for _, line := range block.lines[1:] {
+				newIndentation := countLeadingSpaces(line)
+				if newIndentation <= argIndentation || argIndentation == -1 {
+					header := argHeaderRegex.FindString(line)
+					if len(header) > 0 {
+						arg := argRegex.FindString(header)
+						info.args[arg] = true
+						argIndentation = newIndentation
+					}
+				}
+			}
+		case "Returns:":
+			if len(block.lines) > 1 {
+				info.returns = true
+			}
+		}
+	}
+	return info
+}
+
+func getParamName(param build.Expr) string {
+	switch param := param.(type) {
+	case *build.Ident:
+		return param.Name
+	case *build.BinaryExpr:
+		// keyword parameter
+		if ident, ok := param.X.(*build.Ident); ok {
+			return ident.Name
+		}
+	case *build.UnaryExpr:
+		// *args or **kwargs
+		if ident, ok := param.X.(*build.Ident); ok {
+			return param.Op + ident.Name
+		}
+	}
+	return ""
+}
+
+func hasReturnValues(def *build.DefStmt) bool {
+	result := false
+	build.Walk(def, func(expr build.Expr, stack []build.Expr) {
+		ret, ok := expr.(*build.ReturnStmt)
+		if ok && ret.Result != nil {
+			result = true
+		}
+	})
+	return result
+}
+
+func functionDocstringWarning(f *build.File, fix bool) []*Finding {
+	findings := []*Finding{}
+
+	for _, stmt := range f.Stmt {
+		def, ok := stmt.(*build.DefStmt)
+		if !ok {
+			continue
+		}
+		if strings.HasPrefix(def.Name, "_") {
+			// Private functions aren't required to have docstrings
+			continue
+		}
+		if stmtsCount(def.Body) < FunctionLengthDocstringThreshold {
+			// The function is too short
+			continue
+		}
+		doc, ok := getDocstring(def.Body)
+		if ok {
+			// Docstring exists, check for its contents
+			info := parseFunctionDocstring(doc.(*build.StringExpr))
+			start, end := doc.Span()
+
+			if !info.hasHeader {
+				findings = append(findings, makeFinding(f, start, end, "function-docstring",
+					fmt.Sprintf(`The docstring for the function "%s" should start with a one-line summary.`, def.Name), true, nil))
+			}
+			if info.argumentsLine > 0 {
+				argumentsStart := build.Position{
+					Line:     info.argumentsLine,
+					LineRune: start.LineRune,
+				}
+				argumentsEnd := argumentsStart
+				argumentsEnd.LineRune += len("Arguments:")
+				findings = append(findings, makeFinding(f, argumentsStart, argumentsEnd, "function-docstring",
+					`Prefer 'Args:' to 'Arguments:' when documenting function arguments.`, true, nil))
+			}
+
+			for _, param := range def.Params {
+				name := getParamName(param)
+				if !info.args[name] {
+					findings = append(findings, makeFinding(f, start, end, "function-docstring",
+						fmt.Sprintf(`Argument "%s" is not documented.`, name), true, nil))
```

Comment:
Do we have the opposite warnning: an argument is documented, but doesn't exist

Reply:
Added.